src/main/java/org/threadly/concurrent/PriorityScheduler.java

```diff
@@ -763,232 +639,543 @@ protected void addToExecuteQueue(OneTimeTaskWrapper task) {
+      ArgumentVerifier.assertGreaterThanZero(corePoolSize, "corePoolSize");
+      if (maxPoolSize < corePoolSize) {
+        throw new IllegalArgumentException("maxPoolSize must be >= corePoolSize");
+      }
+      if (threadFactory == null) {
+        threadFactory = new ConfigurableThreadFactory(PriorityScheduler.class.getSimpleName() + "-", true);
+      }
+      
+      poolSizeChangeLock = new Object();
+      workersLock = new Object();
+      availableWorkers = new ArrayDeque<Worker>(corePoolSize);
+      waitingForWorkerCount = 0;
+      currentPoolSize = 0;
+      
+      //calls to verify and set values
+      setKeepAliveTime(keepAliveTimeInMs);
+      setMaxWaitForLowPriority(maxWaitForLowPriorityInMs);
+      
+      this.threadFactory = threadFactory;
+      this.corePoolSize = corePoolSize;
+      this.maxPoolSize = maxPoolSize;
+      this.keepAliveTimeInMs = keepAliveTimeInMs;
+      this.allowCorePoolTimeout = false;
+      shutdownStarted = new AtomicBoolean(false);
+      shutdownFinishing = false;
+      lastHighDelayMillis = 0;
+    }
+
+    /**
+     * Checks if the shutdown has started by an invocation of {@link #startShutdown()}.
+     * 
+     * @return {@code true} if the shutdown has started
+     */
+    public boolean isShutdownStarted() {
+      return shutdownStarted.get();
+    }
+
+    /**
+     * Will start the shutdown of the worker pool.
+     * 
+     * @return {@code true} if this call initiates the shutdown, {@code false} if the shutdown has already started
+     */
+    public boolean startShutdown() {
+      return ! shutdownStarted.getAndSet(true);
+    }
+  
+    /**
+     * Check weather the shutdown process is finished.  In order for the shutdown to finish 
+     * {@link #finishShutdown()} must have been invoked. 
+     * 
+     * @return {@code true} if the scheduler is finishing its shutdown
+     */
+    public boolean isShutdownFinished() {
+      return shutdownFinishing;
+    }
+
+    /**
+     * Finishes the shutdown of the worker pool.  This will ensure all finishing workers are 
+     * killed.
+     */
+    public void finishShutdown() {
+      shutdownFinishing = true;
+      
+      synchronized (workersLock) {
+        Iterator<Worker> it = availableWorkers.iterator();
+        while (it.hasNext()) {
+          Worker w = it.next();
+          it.remove();
+          killWorker(w);
         }
         
-        if (waitTime == Long.MAX_VALUE) {  // prevent overflow
-          workersLock.wait();
-        } else {
-          long elapsedTime = now - startTime;
-          waitTime = maxWaitTimeInMs - elapsedTime;
-          if (waitTime > 0) {
-            workersLock.wait(waitTime);
-          }
-        }
+        // we notify all in case some are waiting for shutdown
+        workersLock.notifyAll();
       }
+    }
+
+    /**
+     * Getter for the current set core worker pool size.
+     * 
+     * @return current core pool size
+     */
+    public int getCorePoolSize() {
```

Comment:
These functions obviously do nothing, they just are to make the code cleaner compared to accessing the protected variables directly.  In the other implementations I did not make functions like these.

What way do you like better?